src/stratis/run.rs

```diff
@@ -4,61 +4,80 @@
 
 //! Main loop
 
-use std::{cell::RefCell, os::unix::io::AsRawFd, rc::Rc};
+use std::{
+    os::unix::io::AsRawFd,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
 
-use nix::sys::signalfd::{signal, SfdFlags, SigSet, SignalFd};
+use async_std::task::block_on;
+use nix::poll::{poll, PollFd, PollFlags};
+use tokio::{
+    runtime::Runtime,
+    select, signal,
+    stream::StreamExt,
+    sync::{
+        mpsc::{channel, Sender},
+        Mutex,
+    },
+    task,
+};
 
 use crate::{
-    engine::{Engine, SimEngine, StratEngine},
+    engine::{Engine, SimEngine, StratEngine, UdevEngineEvent},
     stratis::{
-        dbus_support::MaybeDbusSupport,
-        errors::{StratisError, StratisResult},
-        stratis::VERSION,
+        dm::DmFd, errors::StratisResult, ipc_support::setup, stratis::VERSION,
         udev_monitor::UdevMonitor,
     },
 };
 
-// Process any pending signals, return true if SIGINT received.
-// Return an error if there was an error reading the signal.
-fn process_signal(sfd: &mut SignalFd) -> StratisResult<bool> {
-    match sfd.read_signal() {
-        // This is an unsafe conversion, but in this context that is
-        // mostly harmless. A negative converted value, which is
-        // virtually impossible, will not match any of the masked
-        // values, and stratisd will panic and exit.
-        Ok(Some(sig)) => match sig.ssi_signo as i32 {
-            nix::libc::SIGINT => {
-                info!("SIGINT received, exiting");
-                Ok(true)
+fn udev_thread(sender: Sender<UdevEngineEvent>, should_exit: Arc<AtomicBool>) -> StratisResult<()> {
+    let context = libudev::Context::new()?;
+    let mut udev = UdevMonitor::create(&context)?;
+
+    let mut pollers = [PollFd::new(udev.as_raw_fd(), PollFlags::POLLIN)];
+    loop {
+        match poll(&mut pollers, 100)? {
+            0 => {
+                if should_exit.load(Ordering::Relaxed) {
```

Comment:
Does it make more sense to check ```should_exit``` unconditionally, regardless of return value of ```poll()```?

Reply:
I'll explain my original thought process here and we can discuss what the best way forward is. In the case of a poll value of 0, this means no event was detected and the poll timed out. My design is meant to process all data available to stratisd prior to exiting. Now in the case of the engine, I think this is absolutely necessary so that we don't kill an operation part way through its execution. However, it may be less useful in the case of udev given that the udev database is outside of the engine and we are not responsible for udev state consistency like we are for our own engine. What are your thoughts? Do you think this is unnecessarily cautious? It may honestly be sufficient to exit as long as we're not in the middle of an operation which checking `should_exit` unconditionally would still satisfy.